File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
    vector<int> dist(n, INT_MAX);
    dist[src] = 0;
    
    // Run for k+1 iterations (k stops = k+1 flights)
    for (int i = 0; i <= k; i++) {
        vector<int> temp = dist; // Use previous iteration's distances
        
        for (auto& flight : flights) {
            int from = flight[0];
            int to = flight[1];
            int price = flight[2];
            
            if (dist[from] != INT_MAX) {
                temp[to] = min(temp[to], dist[from] + price);
            }
        }
        
        dist = temp;
    }
    
    return dist[dst] == INT_MAX ? -1 : dist[dst];
}
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        int[][] dist = new int[n][k+2];  // CHANGE 1: k+2 instead of k+1
        for(int[] row : dist){
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> a[0] - b[0]); 

        List<List<int[]>> adj = new ArrayList<>(); 
        for(int i = 0; i < n; i++) adj.add(new ArrayList<>()); 
        for(int[] flight : flights){
            int from = flight[0], to = flight[1], price = flight[2]; 
            adj.get(from).add(new int[]{to,price}); 
        }

        dist[src][k+1] = 0; 
        pq.offer(new int[]{0,src,k+1}); 

        while(!pq.isEmpty()){
            int[] curr = pq.poll(); 

            int currDist = curr[0], node = curr[1], leftDist = curr[2];
            
            // CHANGE 2: Add early return when destination reached
            if(node == dst) return currDist;

            for(int[] next : adj.get(node)){
                int nextNode = next[0]; 
                int weight = next[1]; 
                // CHANGE 3: Use leftDist-1 for comparison
                if(leftDist > 0 && dist[nextNode][leftDist-1] > currDist + weight){
                    dist[nextNode][leftDist-1] = currDist + weight; 
                    pq.offer(new int[]{dist[nextNode][leftDist-1], nextNode, leftDist-1}); 
                }
            }
        }

        return -1;  // CHANGE 4: Simple return -1
    }
}